scripts/provision_layers.py

```diff
@@ -0,0 +1,73 @@
+from argparse import ArgumentParser
+import logging
+from pathlib import Path
+import subprocess
+import sys
+from zipfile import ZipFile
+
+import boto3
+
+from azul import config
+from azul.logging import configure_script_logging
+
+log = logging.getLogger(__name__)
+
+
+class LayersProvisioner:
+    layer_dir = Path(config.project_root) / 'lambdas' / 'layer'
+    out_dir = layer_dir / '.chalice' / 'terraform'
+
+    def __init__(self):
+        self.client = boto3.client('s3')
+
+    def provision_layer(self, bucket, key):
+        log.info('Provisioning layer in bucket: %s with key %s', bucket, key)
+        if self.update_required(bucket, key):
+            log.info('Updating layer.zip')
+            self.update_layer(bucket, key)
+        else:
+            log.info('No update necessary')
+
+    def update_required(self, bucket, key) -> bool:
+        log.info('Checking if layer.zip needs to be updated')
+        try:
+            # Since the object key is content-addressable, just checking for
+            # the file's presence is sufficient
+            self.client.head_object(Bucket=bucket, Key=key)
+        except self.client.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == '404':
+                return True
+        else:
+            return False
+
+    def update_layer(self, bucket, key):
+        input_zip = self.out_dir / 'deployment.zip'
+        layer_zip = self.out_dir / 'layer.zip'
+        self.build_layer_zip(layer_zip, input_zip)
+        self.client.upload_file(str(layer_zip), bucket, key)
+
+    def build_layer_zip(self, destination, input_zip):
+        command = f'chalice package {self.out_dir}'
+        log.info('Running: %s', command)
+        subprocess.check_call(command.split(), cwd=self.layer_dir)
+        log.info('Packaging %s', destination)
+        with ZipFile(input_zip, 'r') as deployment_zip:
+            with ZipFile(destination, 'w') as layer_zip:
+                for file in deployment_zip.namelist():
+                    if file != 'app.py':
+                        data = deployment_zip.read(file)
```

Comment:
Can't we stream files between the two archives? ZipFile.open() can return a writeable handle? If this is significantly more code, don't bother, the complexity isn't worth it.